File: src/com/aslan/module/cipher/keys/StreamKey_V1.js
import MixKey from "./MixKey";
import {S1} from "./box";
import {S2} from "./box";

export default class StreamKey extends MixKey {
	indices = null;

	init(cipherInfo, algorithmInfo) {
		super.init(cipherInfo, algorithmInfo);
		this.init_key(cipherInfo.key);
		this.update_key();
		this.update_key();
	}

	update() {
		let indices = this.indices;
		let R = this.R, H = this.H, N = this.N;
		let x = 0, v = R - 1, h = H - 1;
		for (let i = 0; i < R; i++, v--) {
			x = indices[i] & h;
			let U = this.K[i];
			let V = this.K[v];
			for (let y = H; y < N;) {
				U[y] = (S1[(U[y] ^ V[y])] ^ U[x]);
				U[x] = (S2[(U[x] ^ V[x])] ^ U[y]);
				x = (++x) & h, y++;
				U[y] = (S1[(U[y] ^ V[y])] ^ U[x]);
				U[x] = (S2[(U[x] ^ V[x])] ^ U[y]);
				x = (++x) & h, y++;
				U[y] = (S1[(U[y] ^ V[y])] ^ U[x]);
				U[x] = (S2[(U[x] ^ V[x])] ^ U[y]);
				x = (++x) & h, y++;
				U[y] = (S1[(U[y] ^ V[y])] ^ U[x]);
				U[x] = (S2[(U[x] ^ V[x])] ^ U[y]);
				x = (++x) & h, y++;
				U[y] = (S1[(U[y] ^ V[y])] ^ U[x]);
				U[x] = (S2[(U[x] ^ V[x])] ^ U[y]);
				x = (++x) & h, y++;
				U[y] = (S1[(U[y] ^ V[y])] ^ U[x]);
				U[x] = (S2[(U[x] ^ V[x])] ^ U[y]);
				x = (++x) & h, y++;
				U[y] = (S1[(U[y] ^ V[y])] ^ U[x]);
				U[x] = (S2[(U[x] ^ V[x])] ^ U[y]);
				x = (++x) & h, y++;
				U[y] = (S1[(U[y] ^ V[y])] ^ U[x]);
				U[x] = (S2[(U[x] ^ V[x])] ^ U[y]);
				x = (++x) & h, y++;
			}
			indices[i] = (indices[i] << 1);
			indices[i] = (indices[i]) > H ? 1 : indices[i];
		}
		return this.K;
	}

	alloc(algorithmInfo) {
		super.alloc(algorithmInfo);
		this.indices = new Uint32Array(algorithmInfo.R);
	}

	init_key(key) {
		super.init_key(key);
		this.init_indices(this.R, this.N);
	}

	init_indices(round, N) {
		this.indices[0] = N;
		for (let i = 1; i <= this.R; i++) {
			this.indices[i] = 1 << i - 1;
		}
	}

	get version() {
		return 1;
	}

}

File: src/com/aslan/module/cipher/keys/StreamKey_V1.java
package com.aslan.module.cipher.keys;

import com.aslan.module.cipher.AlgorithmInfo;
import com.aslan.module.cipher.CipherInfo;

import static com.aslan.module.cipher.keys.Box.*;

public class StreamKey_V1 extends MixKey {
    int[] indices = null;

    public void init(CipherInfo cipherInfo, AlgorithmInfo algorithmInfo) {
        super.init(cipherInfo, algorithmInfo);
        init_key(cipherInfo.key);
        update_key();
        update_key();
    }

    public byte[][] update() {
        int x = 0, v = R - 1, h = H - 1;
        for (int i = 0; i < R; i++, v--) {
            x = indices[i] & h;
            byte[] U = K[i];
            byte[] V = K[v];
            for (int y = H; y < N; ) {
                U[y] = (byte) (S1[(U[y] ^ V[y])] ^ U[x]);
                U[x] = (byte) (S2[(U[x] ^ V[x])] ^ U[y]);
                x = (++x) & h;
                y++;

            }
            indices[i] = (indices[i] << 1);
            indices[i] = (indices[i]) > H ? 1 : indices[i];
        }
        return K;
    }

    protected void alloc(AlgorithmInfo algorithmInfo) {
        super.alloc(algorithmInfo);
        indices = new int[algorithmInfo.R];
    }

    protected void init_key(byte[] key) {
        super.init_key(key);
        init_indices(N);
    }

    protected void init_indices(int N) {
        indices[0] = N;
        for (int i = 1; i <= R; i++) {
            indices[i] = 1 << i - 1;
        }
    }

    @Override
    public int level() {
        return 2;
    }

    @Override
    public int version() {
        return 1;
    }
}